### PR TITLE
add save_last_n_optim_states flag

### PR DESCRIPTION
**Pull request by dimapihtar**
# What does this PR do ?

1. Adds save_last_n_optim_states to save optimizer states only for last n checkpoints. This approach allows to save the disk space. Checkpoint file without optimizer states weighs about 7 times less
2. Now it's also possible to load checkpoint without optimizer states for the inference.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* To save optimizer states for last 2 checkpoints, set this config value:

```python
exp_manager.checkpoint_callback_params.save_last_n_optim_states=2 
```
`save_last_n_optim_states=-1` is set by default. It means that optimizer states will be saved for each checkpoint. In order to drop optimizer states for each training checkpoint, you can set `save_last_n_optim_states=0`.

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by ericharper** [comment]
closes #9967

File: .github/workflows/cicd-main.yml [hunk 1]
```diff
@@ -2776,6 +2776,43 @@ jobs:
...
+        trainer.log_every_n_steps=1 \
+        trainer.val_check_interval=2 \
+        trainer.limit_val_batches=2 \
+        trainer.accumulate_grad_batches=1 \
+        trainer.max_steps=6 \
+        trainer.precision=bf16 \
+        trainer.gradient_clip_val=1.0 \
+        exp_manager.exp_dir=examples/nlp/language_modeling/gpt_pretrain_results \
+        model.save_last_n_optim_states=1 \
+        model.dist_ckpt_format='torch_dist' \
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:2797]:
```suggestion
        model.dist_ckpt_format="torch_dist" \
```

**Review by ko3n1g — CHANGES_REQUESTED** [review]
one small required change to make the gpu go burr burr and then it should be fine

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 2]
```diff
@@ -410,6 +412,9 @@ def save_checkpoint(
             if self.is_global_zero or app_state.data_parallel_rank == 0:
                 self.checkpoint_io.save_checkpoint(checkpoint, filepath, storage_options=storage_options)
 
+        if self.save_last_n_optim_states >= 0 and '-last' in filepath:
```
Comment (mikolajblaz) [inline nemo/collections/nlp/parts/nlp_overrides.py:415]: This creates a big entanglement in the code.

`-last` is very specific to ModelCheckpoint callbacks, the strategy shouldn't know about stuff like this and it shouldn't happen during `save_checkpoint` call

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 3]
```diff
@@ -584,6 +589,51 @@ def _integrate_original_checkpoint_data(self, checkpoint: Dict[str, Any]) -> Dic
 
         return checkpoint
 
+    def _drop_optimizer_states(self, filepath: Union[str, Path], storage_options: Optional[Any] = None) -> None:
```
Comment (mikolajblaz) [inline nemo/collections/nlp/parts/nlp_overrides.py:592]: most of the code below conceptually fits into ModelCheckpoint class (except for state dict construction, see [here](https://github.com/NVIDIA/NeMo/pull/10241/files#r1736080010))

I would suggest moving it all there except for this part which I guess can stay at `save_checkpoint` and we can just decide whether we want to include optimizer state in `storage_options`
 ```
            checkpoint['optimizer_states'] = [None]
            checkpoint['state_dict'] = OrderedDict([])
            checkpoint['sharded_state_dict'] = self.lightning_module.sharded_state_dict()

            # Save the checkpoint without optimizer states
            self.checkpoint_io.save_checkpoint(checkpoint, checkpoint_path, storage_options=storage_options)

            if HAVE_MODELOPT and hasattr(self.lightning_module, "get_model_module_list"):
                save_sharded_modelopt_state(
                    self.lightning_module.get_model_module_list(),
                    checkpoint_path,
                    self.unwrapped_checkpoint_io.save_sharded_strategy,
                    prefix="model.",
                )
```

**Review by mikolajblaz — COMMENTED** [review]
I think this approach is correct, although quite inefficient (after a few saves, we will load and save old ckpt on **each** new ckpt save).

For now I think it's fine for users who need this functionality, long term we can separate the model and optimizer state inside the checkpoint so that it's easy to remove the opt state without load/save.

**Comment by mikolajblaz** [comment]
> I think this approach is correct, although quite inefficient (after a few saves, we will load and save old ckpt on **each** new ckpt save).
> 
> For now I think it's fine for users who need this functionality, long term we can separate the model and optimizer state inside the checkpoint so that it's easy to remove the opt state without load/save.

For a more efficient implementation of dropping optimizer state, I see 2 alternatives:
- One is NeMo specific and described here: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/issues/578#note_24602416
- second option is adding a capability to separate different parts of the checkpoint **within** the `torch_dist` format so that it's easy to drop one of them in-place. This would be implemented inside MCore and NeMo would just specify that `optimizier_states` is the part to separate during saving and drop during cleanup.

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 4]
```diff
@@ -584,6 +589,51 @@ def _integrate_original_checkpoint_data(self, checkpoint: Dict[str, Any]) -> Dic
...
+        # Drop optimizer states
+        checkpoint_index = len(checkpoints) - self.save_last_n_optim_states - 1
+        if len(checkpoints) > self.save_last_n_optim_states:
+            checkpoint_path = checkpoints[checkpoint_index]
+
+            logging.info(f"Loading '{checkpoint_path}' checkpoint to drop optimizer states...")
+            checkpoint = self.load_checkpoint(checkpoint_path=checkpoint_path)
+
+            # Remove the checkpoint version with optimizer states
+            self.remove_checkpoint(checkpoint_path)
```
[the remark this answers is not in the document]
Comment (dimapihtar) [reply nemo/collections/nlp/parts/nlp_overrides.py:605]: @mikolajblaz but in this case we'll get an error that checkpoint exitst and not empty or should we change the name for checkpoint w/o optimizer states?

[the remark this answers is not in the document]
Comment (mikolajblaz) [reply nemo/collections/nlp/parts/nlp_overrides.py:605]: We probably have to save to a temp dir and rename.
Imagine the following case: at iter 1000 we have 3 checkpoints from iters 800, 900, 1000 and we want to remove the optimizer state from iter 80. If during removal we have an error we lose the iter 800 checkpoint forever (we can't even retrain the model for a while) and it could be e.g. checkpoint with Top1 val_loss

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 5]
```diff
@@ -545,7 +556,9 @@ def load_checkpoint(self, checkpoint_path: Union[str, Path]) -> Dict[str, Any]:
 
             # after dist_checkpointing.load, sharded tensors will be replaced with tensors
             checkpoint['state_dict'] = sharded_state_dict
-            checkpoint['optimizer_states'] = [self.optimizer_sharded_state_dict(is_loading=True)]
+            # Check whether to load optim states
+            if map_location.get("load_optim_states", True):
```
Comment (mikolajblaz) [inline nemo/collections/nlp/parts/nlp_overrides.py:560]: We shouldn't use `map_location` for that purpose, it's not even in the base strategy interface?

For the record, this approach will work even if we don't restrict optimizer loading, because we won't save the optimizer state. But restricted load will be way faster because optimizer state is the majority of the checkpoint.

I would suggest adding a new argument to `load_checkpoint`, e.g. `load_optimizer_state: bool = True`

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 6]
```diff
@@ -391,7 +394,12 @@ def save_checkpoint(
             sharded_optim_state = self.optimizer_sharded_state_dict(
                 unsharded_optim_state=checkpoint['optimizer_states'][0]
             )
-            checkpoint['optimizer_states'] = [sharded_optim_state]
+
+            # Check whether to save optim states
+            if storage_options.get('drop_optim_states', False):
```
Comment (mikolajblaz) [inline nemo/collections/nlp/parts/nlp_overrides.py:399]: Even though these are different strategies, maybe let's decide on a single name to reduce confusion, see parallel PR for NeMo 2.0: https://github.com/NVIDIA/NeMo/pull/10241/files#diff-78f81f4094cfea056c177e87c0d527b9ce27cee11813138e5a2a69370b922c19R610
I'm OK with both.

cc @ashors1

Comment (dimapihtar) [reply nemo/collections/nlp/parts/nlp_overrides.py:399]: okay, I'm fine with the name @ashors1 used. Not a big deal I think.

**Review by mikolajblaz — APPROVED** [review]

File: nemo/utils/callbacks/nemo_model_checkpoint.py [hunk 7]
```diff
@@ -198,6 +201,10 @@ def setup(self, trainer, pl_module, stage: str) -> None:
         self.last_model_path = trainer.strategy.broadcast(self.last_model_path)
 
     def on_save_checkpoint(self, trainer, pl_module, checkpoint):
+        # Load the correct state_dict for current checkpoint.
+        # Temporary solution.
+        if self.save_last_n_optim_states >= 0:
+            self._load_current_state_dict(trainer, checkpoint)
```
Comment (mikolajblaz) [inline nemo/utils/callbacks/nemo_model_checkpoint.py:207]: why is this in `on_save_checkpoint`?

Comment (mikolajblaz) [reply nemo/utils/callbacks/nemo_model_checkpoint.py:207]: Even if for some reason we need to put it here, let's make sure the `._load_current_state_dict` is not executed for "regular" saves

File: nemo/utils/callbacks/nemo_model_checkpoint.py [hunk 8]
```diff
@@ -361,6 +364,78 @@ def _ema_callback(self, trainer: 'pytorch_lightning.Trainer') -> Optional[EMA]:
...
+                trainer.strategy.remove_checkpoint(checkpoint_path)
+                shutil.move(f"{checkpoint_path}{suffix}", checkpoint_path)
+
+            if torch.distributed.is_initialized():
+                torch.distributed.barrier()
+
+            # Load the correct state_dict for current checkpoint.
+            # Temporary solution.
+            checkpoint = trainer.strategy.load_checkpoint(
+                checkpoint_path=ckpt_to_dir(filepath), load_optimizer_states=True
```
Comment (mikolajblaz) [inline nemo/utils/callbacks/nemo_model_checkpoint.py:404]: Did you test with load_optimizer_states=False maybe? It will be way faster

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 9]
```diff
@@ -391,7 +391,14 @@ def save_checkpoint(
             sharded_optim_state = self.optimizer_sharded_state_dict(
                 unsharded_optim_state=checkpoint['optimizer_states'][0]
             )
-            checkpoint['optimizer_states'] = [sharded_optim_state]
+
+            # Check whether to save optim states
```
Comment (mikolajblaz) [inline nemo/collections/nlp/parts/nlp_overrides.py:395]: since we are not modifying storage_options maybe let's not turn it into a dict (just check if it's not None line below)

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 10]
```diff
@@ -470,6 +477,8 @@ def _check_param_groups_mismatch(self, checkpoint_path: Union[str, Path], sharde
         # @akoumparouli: check if it contains an mcore dist opt
         if common_state_dict.get('optimizer_states', [{}])[0].get('param_groups', None) is None:
             return False
+        if sharded_state_dict.get('optimizer_states') is None:
```
Comment (mikolajblaz) [inline nemo/collections/nlp/parts/nlp_overrides.py:480]: [Minor] If you put it before the previous check, you can simplify the `common_state_dict.get('optimizer_states', [{}])` to `common_state_dict['optimizer_states']`

File: nemo/utils/callbacks/nemo_model_checkpoint.py (hunk 8 above)
Comment (dimapihtar) [reply nemo/utils/callbacks/nemo_model_checkpoint.py:404]: yes, it works fine with load_optimizer_states=False

**Review by mikolajblaz — APPROVED** [review]

**Review by pablo-garay — APPROVED** [review]
